controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java

```diff
@@ -1168,48 +1171,65 @@ public void testDeployComplicatedDeploymentSpec() {
         assertEquals(Change.empty(), app2.instance().change());
         assertEquals(Change.empty(), app3.instance().change());
 
-        // Upgrade instance 1; a failure in any instance allows an application change to accompany the upgrade.
+        // Upgrade instance 1; upgrade rolls out first, with revision following.
         // The new platform won't roll out to the conservative instance until the normal one is upgraded.
-        app2.failDeployment(systemTest);
         app1.submit(applicationPackage);
         assertEquals(Change.of(version).with(app1.application().latestVersion().get()), app1.instance().change());
+        // Upgrade platform.
         app2.runJob(systemTest);
-        app1.jobAborted(stagingTest)
-            .runJob(stagingTest)
+        app1.runJob(stagingTest)
             .runJob(productionUsWest1)
             .runJob(productionUsEast3);
-        app1.runJob(stagingTest);   // Tests with only the outstanding application change.
-        app2.runJob(systemTest);    // Tests with only the outstanding application change.
+        // Upgrade revision
+        tester.clock().advance(Duration.ofSeconds(1)); // Ensure we see revision as rolling after upgrade.
+        app2.runJob(systemTest);        // R
+        app1.runJob(stagingTest)        // R
+            .runJob(productionUsWest1); // R
+            // productionUsEast3 won't change revision before its production test has completed for the upgrade, which is one of the last jobs!
         tester.clock().advance(Duration.ofHours(2));
         app1.runJob(productionEuWest1);
         tester.clock().advance(Duration.ofHours(1));
         app1.runJob(productionAwsUsEast1a);
-        tester.triggerJobs();
         app1.runJob(testAwsUsEast1a);
+        tester.clock().advance(Duration.ofSeconds(1));
+        app1.runJob(productionAwsUsEast1a); // R
+        app1.runJob(testAwsUsEast1a);       // R
         app1.runJob(productionApNortheast2);
         app1.runJob(productionApNortheast1);
         tester.clock().advance(Duration.ofHours(1));
         app1.runJob(testApNortheast1);
         app1.runJob(testApNortheast2);
+        app1.runJob(productionApNortheast2); // R
+        app1.runJob(productionApNortheast1); // R
         app1.runJob(testUsEast3);
         app1.runJob(productionApSoutheast1);
+        tester.clock().advance(Duration.ofSeconds(1));
+        app1.runJob(productionUsEast3);      // R
+        tester.clock().advance(Duration.ofHours(2));
+        app1.runJob(productionEuWest1);      // R
+        tester.clock().advance(Duration.ofMinutes(330));
+        app1.runJob(testApNortheast1);       // R
+        app1.runJob(testApNortheast2);       // R
+        app1.runJob(testUsEast3);            // R
```

Comment:
Should probably make a job in between production-us-east-3 and this fail, so we see that the upgrade-only run for test-us-east-3 is abandoned, and the revision is deployed earlier, to potentially fix that upgrade failure.